tests/library/chromium/chromium.spec.ts

```diff
@@ -30,6 +30,233 @@ test('should create a worker from a service worker', async ({ page, server }) =>
   expect(await worker.evaluate(() => self.toString())).toBe('[object ServiceWorkerGlobalScope]');
 });
 
+test('should create a worker from service worker with noop routing', async ({ context, page, server }) => {
+  await context.route('**', route => route.continue());
+  const [worker] = await Promise.all([
+    page.context().waitForEvent('serviceworker'),
+    page.goto(server.PREFIX + '/serviceworkers/empty/sw.html')
+  ]);
+  expect(await worker.evaluate(() => self.toString())).toBe('[object ServiceWorkerGlobalScope]');
+});
+
+test('isServiceWorker() and serviceWorker() work', async ({ context, page, server }) => {
+  {
+    const [worker, html, main, inWorker] = await Promise.all([
+      context.waitForEvent('serviceworker'),
+      context.waitForEvent('request', r => r.url().endsWith('/sw.html')),
+      context.waitForEvent('request', r => r.url().endsWith('/sw.js')),
+      context.waitForEvent('request', r => r.url().endsWith('/request-from-within-worker.txt')),
+      page.goto(server.PREFIX + '/serviceworkers/fetch/sw.html')
+    ]);
+    const [inner] = await Promise.all([
+      context.waitForEvent('request', r => r.url().endsWith('/inner.txt')),
+      page.evaluate(() => fetch('/inner.txt')),
+    ]);
+    expect(html.frame()).toBeTruthy();
+    expect(html.isServiceWorkerRequest()).toBe(false);
+    expect(html.serviceWorker()).toBe(null);
+
+    expect(main.frame).toThrow();
+    expect(main.isServiceWorkerRequest()).toBe(true);
+    expect(main.serviceWorker()).toBe(worker);
+
+    expect(inner.frame()).toBeTruthy();
+    expect(inner.isServiceWorkerRequest()).toBe(false);
+    expect(inner.serviceWorker()).toBe(null);
+
+    expect(inWorker.frame).toThrow();
+    expect(inWorker.isServiceWorkerRequest()).toBe(true);
+    expect(inWorker.serviceWorker()).toBe(worker);
+  }
+  await page.evaluate(() => window['activationPromise']);
+  {
+    const [inner] = await Promise.all([
+      context.waitForEvent('request', r => r.url().endsWith('/inner.txt')),
+      page.evaluate(() => fetch('/inner.txt')),
+    ]);
+    // FIXME: This doesn't seem to be the correct assetion.
+    // Ensure the reques is actually intercepted by SW.
+    expect(inner.isServiceWorkerRequest()).toBe(false);
+    expect(inner.serviceWorker()).toBe(null);
```

Comment:
There's a bug somewhere! This is wrong. It should be considered service worker. Interestingly, if you add `context.route('**', (route) => route.continue());` to the top of the test, it passes. (Having the no-op context route does enable more CDP events.)

Reply:
For SW with `fetch` listener, it looks like we are firing two requestWillBeSentEvents. One on Page and one on SW.